### PR TITLE
Implemented completion, improvements to hovering and diagnostic fetching, caching support, bugfixes

### PR DESCRIPTION
**Pull request by wz1000**

### CODE REVIEW

File: app/MainHie.hs [hunk 1]
```diff
@@ -153,10 +153,10 @@ run opts = do
       Just err -> error (pdeErrorMsg err)
       Nothing -> return ()
 
-    -- let vomitOptions = GM.defaultOptions { GM.optOutput = oo { GM.ooptLogLevel = GM.GmVomit}}
-    --     oo = GM.optOutput GM.defaultOptions
-    -- let ghcModOptions = vomitOptions { GM.optGhcUserOptions = ["-Wall"]  }
-    let ghcModOptions = GM.defaultOptions { GM.optGhcUserOptions = ["-Wall"]  }
+    let vomitOptions = GM.defaultOptions { GM.optOutput = oo { GM.ooptLogLevel = GM.GmVomit}}
```
Comment (alanz) [inline app/MainHie.hs]: Probably best to disable this before merging into master.

And best of all to make it subject to a flag

File: haskell-ide-engine.cabal [hunk 2]
```diff
@@ -130,8 +130,8 @@ executable hie
                      , unordered-containers
                      , vinyl
   ghc-options:         -threaded -rtsopts -with-rtsopts=-N -Wall
-  if flag(pedantic)
-     ghc-options:      -Werror
+--  if flag(pedantic)
```
Comment (alanz) [inline haskell-ide-engine.cabal]: And this needs to be reinstated

Comment (wz1000) [inline haskell-ide-engine.cabal]: hmm, but it breaks ghc-mod.

File: app/MainHie.hs (hunk 1 above)
Comment (wz1000) [inline app/MainHie.hs]: ok

**Pull request merged**